pilot/pkg/config/kube/ior/route.go

```diff
@@ -387,64 +323,24 @@ func (r *route) handleEvent(event model.Event, cfg config.Config) error {
 
 	switch event {
 	case model.EventAdd:
-		return r.handleAdd(cfg)
+		return r.onGatewayAdded(cfg)
 
 	case model.EventUpdate:
-		if err = r.verifyResourceVersions(cfg); err != nil {
-			return err
-		}
-
-		var result *multierror.Error
-		result = multierror.Append(result, r.handleDel(cfg))
-		result = multierror.Append(result, r.handleAdd(cfg))
-		return result.ErrorOrNil()
+		return r.onGatewayUpdated(cfg)
 
 	case model.EventDelete:
-		return r.handleDel(cfg)
+		return r.onGatewayRemoved(cfg)
 	}
```

Comment:
Are we sure that we always get the correct event here? For example, is it possible to get an ADD event when the Gateway already exists? 

I'm concerned that IOR will continue to be edge-triggered, whereas controllers in Kubernetes should be level-triggered. IMO all three events should invoke the same function ("reconcileGateway"). This function should be idempotent and only consider the current state of the Gateway object (not the change in state as it does now). The controller would be much more robust.

Reply:
This is the [queue](https://github.com/maistra/istio/blob/maistra-2.2/pkg/queue/instance.go#L100) implementation IOR relying on. The queue is populated via an [informer](https://github.com/maistra/istio/blob/maistra-2.2/pilot/pkg/config/kube/crdclient/cache_handler.go#L86) of `Gateway` kind. The queue will guarantee the order of the event for us.

For an ADD event, this is triggered after the Gateway gets recreated. The error message showed in the log is related to that  house keeping map in the IOR for us to handle deletion. The error tells that the CREATE event is not there yet but it sees an UPDATE.

I agree on the idempotent. I am making some changes, so it will discard event if it is outdated.